### PR TITLE
[WIP] Fix edge cases in TopPLogitsWarper when top_p equals 0 or 1

### PR DESCRIPTION
**Pull request by NinedayWang**
# What does this PR do?

This PR fixes edge cases in TopPLogitsWarper when top_p equals 0 or 1. The following are specific contributions:
1. Make the code consistent with the `ValueError` error message.
2. Fix the `nan` problem. In the original PyTorch implementation, if `do_sample=True` and `top_p=0`, then each position in logits will be set to `-float("Inf")`, and then the `nan` problem will be encountered after softmax.
3. Make the three frameworks (PyTorch/TF/FLAX) have the same behavior. In the original implementation, when `do_sample=True` and `top_p=0`, the PyTorch framework would encounter `nan` errors, while TF and FLAX would keep the highest scoring token due to their right shift operations.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@gante @patrickvonplaten

### CODE REVIEW

**Comment by gante** [comment]
Hi @NinedayWang 👋 

The edge cases are intentionally breaking -- `top_p=0.0` means in theory that no token can be sampled (and in practice, that they have the same odds), while `top_p=1.0` is equivalent to not having the `top_p` operation.

I struggle to see a use case for these changes, but I'm open to suggestions :)

**Comment by gante** [comment]
@NinedayWang thank you for elaborating -- that makes sense! :) 

I'd like to suggest two modifications, to then merge the PR:
1. Regarding making the check on `top_p` more strict: because our library is used in production, we must avoid breaking changes whenever we can (which would be the case). Instead, can we raise a warning when `top_p` is `0.0` (degenerated to argmax token selection) or `1.0` (redundant operation)?
2. Regarding the case where `top_p=0.0`: we both actually forgot the other argument in the discussion above, `min_tokens_to_keep`, which is not being respected. To fix it in this edge case, changing `if self.min_tokens_to_keep > 1:` to `if self.min_tokens_to_keep > 0:` OR forcing `min_tokens_to_keep` to be strictly positive (and removing this `if`) is enough. This change will also make the edge case return back to a greedy decoding-like behavior, as in TF and FLAX, with no exceptions being thrown.

**Comment by NinedayWang** [comment]
@gante Thanks for your advice! I will work on this :)

**Comment by gante** [comment]
@NinedayWang Are you planning to continue this work? :)

**Comment by NinedayWang** [comment]
> 

Terribly sorry! Other urgent matters delayed this work. I will continue it and update the progress soon!

**Comment by NinedayWang** [comment]
> @NinedayWang Are you planning to continue this work? :)

Terribly sorry! Other urgent matters delayed this work. I will continue it and update the progress soon!

**Comment by gante** [comment]
@NinedayWang no worries, take your time :D
